### PR TITLE
Add tests for license changes while using data streams

### PR DESCRIPTION
**Pull request by lkts**
Depends on #115397.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 1]
```diff
@@ -980,34 +980,55 @@ static Settings aggregateIndexSettings(
 
         final Settings.Builder indexSettingsBuilder = Settings.builder();
         if (sourceMetadata == null) {
-            final Settings templateAndRequestSettings = Settings.builder().put(combinedTemplateSettings).put(request.settings()).build();
+            final Settings.Builder settingsBuilder = Settings.builder().put(combinedTemplateSettings).put(request.settings());
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:983]: The changes to the file can be reverted wow that #115437 has been merged.

**Review by martijnvg — APPROVED** [review]
One comment, LGTM otherwise.

**Review by kkrik-es — APPROVED** [review]
Are we planning to add more tests, where the mode gets implicitly derived based on the index name? I wonder if we can add some sort of randomization, as there are many ways to end up with logsdb and synthetic source.

**Comment by lkts** [comment]
@kkrik-es sorry, i don't follow. Tests here uses implicit values for both index mode and source mode. First data stream we test has name that matches `logs-*-*` and nothing else.

**Comment by kkrik-es** [comment]
> @kkrik-es sorry, i don't follow. Tests here use implicit values for both index mode and source mode. First data stream we test has name that matches `logs-*-*` and nothing else.

Don't we have  "mapping.source.mode": "SYNTHETIC" etc? I think we should have a mix of explicit and implicit settings.

**Comment by lkts** [comment]
There is a data stream `logs-test-regular` which has nothing and another one `logs-test-custom` indeed has source mode configuration.

**Comment by lkts** [comment]
@elasticmachine update branch

**Comment by lkts** [comment]
Re-requesting since i added a bunch of tests.

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/DataStreamLicenceDowngradeIT.java [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    }
+
+    @Override
+    protected void licenseChange() throws IOException {
+        startBasic();
+    }
+
+    @Override
+    protected List<TestCase> cases() {
+        return List.of(new TestCase() {
```
Comment (kkrik-es) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/DataStreamLicenceDowngradeIT.java:28]: Nit: I'd add the test cases as separate subclasses below to fascilitate debugging, and include them in the list here.

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/DataStreamLicenseUpgradeIT.java [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+    protected void applyInitialLicense() {}
+
+    @Override
+    protected void licenseChange() throws IOException {
+        startTrial();
+    }
+
+    @Override
+    protected List<TestCase> cases() {
+        return List.of(new TestCase() {
```
Comment (kkrik-es) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/DataStreamLicenseUpgradeIT.java:26]: Some of the test cases are shared with downgrade tests, with the outcome flipped or adjusted. Maybe they can be futher refactored, with upgrade/downgrade just giving a different pair of initial/final mode.

**Review by kkrik-es — APPROVED** [review]
Well done. I added a few ideas about how to further refactor, they can be applied in a follow-up if applicable.

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**